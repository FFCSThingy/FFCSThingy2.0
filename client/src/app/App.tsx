import React, { useEffect } from 'react';
import {
	BrowserRouter as Router, Switch, Route, Redirect,
} from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';

import ReactGA from 'react-ga';

import Login from '../components/Login';
import Dashboard from '../components/Dashboard';

import { checkAuth } from '../reducers/auth';
import { selectIsAuthenticated } from '../selectors/auth';

ReactGA.initialize('UA-156974674-1');
ReactGA.pageview(window.location.pathname + window.location.search);

const App = () => {
	const dispatch = useDispatch();
	const isAuthenticated = useSelector(selectIsAuthenticated);

	useEffect(() => {
		dispatch(checkAuth());
	// Don't add more dependencies
	// Only needs to run once
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, []);

	return (
		<Router>
			<Switch>
				<Route path="/login">
					{!isAuthenticated
						? <Login />
						: <Redirect to="/" />}
				</Route>

				<Route path="/">
					<Dashboard />
					{/* {isAuthenticated
						? <Dashboard />
						: <Redirect to="/login" />} */}
				</Route>
			</Switch>
		</Router>
	);
};

export default App;
