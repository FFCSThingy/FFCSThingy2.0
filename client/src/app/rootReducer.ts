import { combineReducers } from 'redux';

import miscReducer,
{
	initialState as miscInitState,
	name as miscName,
} from '../reducers/misc';

import courseReducer,
{
	initialState as courseInitState,
	name as courseName,
} from '../reducers/course';

import timetableReducer,
{
	initialState as timetableInitState,
	name as timetableName,
} from '../reducers/timetable';

import curriculumReducer,
{
	initialState as curriculumInitState,
	name as curriculumName,
} from '../reducers/curriculum';

import userReducer,
{
	initialState as userInitState,
	name as userName,
} from '../reducers/user';

import authReducer,
{
	initialState as authInitState,
	name as authName,
} from '../reducers/auth';

const rootReducer = combineReducers({
	misc: miscReducer,
	course: courseReducer,
	timetable: timetableReducer,
	curriculum: curriculumReducer,
	user: userReducer,
	auth: authReducer,
});

export const initialState = {
	[miscName]: miscInitState,
	[courseName]: courseInitState,
	[timetableName]: timetableInitState,
	[curriculumName]: curriculumInitState,
	[userName]: userInitState,
	[authName]: authInitState,
};

export type RootState = ReturnType<typeof rootReducer>;
export default rootReducer;
