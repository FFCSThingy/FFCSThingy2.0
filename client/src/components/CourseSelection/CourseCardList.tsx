import React, { FC } from 'react';
import { CardColumns } from 'react-bootstrap';
import { useSelector } from 'react-redux';

import CourseCard from './CourseCard';

import styles from '../../css/CourseSelectionList.module.scss';

import CourseCardListProps from '../../models/components/CourseSelection/CourseCardList';
import { selectPrereqList, selectSelectedCourse } from '../../selectors/course';
import { selectCompletedCourses } from '../../selectors/user';

const CourseCardList: FC<CourseCardListProps> = ({
	filteredCourseList = {},
}) => {
	const selectedCourse = useSelector(selectSelectedCourse);
	const prereqList = useSelector(selectPrereqList);
	const completedCourses = useSelector(selectCompletedCourses);

	const completedCourseCards = Object.keys(filteredCourseList)
		.filter((code) => completedCourses && completedCourses[code])
		.map((code) => {
			const { title, credits, shortCourseTypes } = filteredCourseList[code];

			return (
				<CourseCard
					code={code}
					title={title}
					credits={credits}
					shortCourseTypes={shortCourseTypes}
					selected={(selectedCourse === code)}
					completed={completedCourses[code]}
					prereqs={prereqList[code]}
					key={`${code}-CourseCard`}
				/>
			);
		});

	const normalCards = Object.keys(filteredCourseList)
		.filter((code) => !completedCourses || (completedCourses && !completedCourses[code]))
		.map((code) => {
			const { title, credits, shortCourseTypes } = filteredCourseList[code];

			return (
				<CourseCard
					code={code}
					title={title}
					credits={credits}
					shortCourseTypes={shortCourseTypes}
					selected={(selectedCourse === code)}
					prereqs={prereqList[code]}
					key={`${code}-CourseCard`}
				/>
			);
		});

	return (
		<CardColumns className={styles.courseList}>
			{normalCards}
			{completedCourseCards}
		</CardColumns>
	);
};

export default CourseCardList;
