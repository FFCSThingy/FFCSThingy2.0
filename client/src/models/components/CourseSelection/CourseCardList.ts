import { CourseList } from '../../data/CourseLists';

export default interface CourseCardListProps {
	filteredCourseList?: CourseList;
}
