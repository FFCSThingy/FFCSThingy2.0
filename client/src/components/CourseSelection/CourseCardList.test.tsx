import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import CourseCardList from './CourseCardList';

import { initialState } from '../../app/rootReducer';
import { CourseList } from '../../models/data/CourseLists';

const mockStore = configureStore([thunk]);

const courseList: CourseList = {
	ZZZ1001: {
		title: 'Sample1',
		credits: 4,
		types: ['ETH', 'ELA', 'EPJ'],
		simpleCourseTypes: ['Lab', 'Theory', 'Project'],
		shortCourseTypes: ['L', 'T', 'P'],
	},
	ZZZ1002: {
		title: 'Sample2',
		credits: 4,
		types: ['ETH', 'ELA', 'EPJ'],
		simpleCourseTypes: ['Lab', 'Theory', 'Project'],
		shortCourseTypes: ['L', 'T', 'P'],
	},
	ZZZ1003: {
		title: 'Sample3',
		credits: 4,
		types: ['ETH', 'ELA', 'EPJ'],
		simpleCourseTypes: ['Lab', 'Theory', 'Project'],
		shortCourseTypes: ['L', 'T', 'P'],
	},
};

describe('component: CourseCardList', () => {
	it('should render without explicit props', () => {
		expect.hasAssertions();

		const store = mockStore(initialState);
		const { asFragment } = render(
			<Provider store={store}>
				<CourseCardList />
			</Provider>,
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render with filtered course list', () => {
		expect.hasAssertions();

		const store = mockStore(initialState);
		const { asFragment } = render(
			<Provider store={store}>
				<CourseCardList
					filteredCourseList={courseList}
				/>
			</Provider>,
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render completed courses if data present', () => {
		expect.hasAssertions();

		const store = mockStore({
			...initialState,
			user: {
				...initialState.user,
				completedCourses: {
					ZZZ1001: 'N',
					ZZZ1003: 'A',
				},
			},
		});
		const { asFragment } = render(
			<Provider store={store}>
				<CourseCardList
					filteredCourseList={courseList}
				/>
			</Provider>,
		);

		expect(asFragment()).toMatchSnapshot();
	});
});
