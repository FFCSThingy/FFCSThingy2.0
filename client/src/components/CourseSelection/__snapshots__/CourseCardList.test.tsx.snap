// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: CourseCardList should render completed courses if data present 1`] = `
<DocumentFragment>
  <div
    class="courseList card-columns"
  >
    <div
      class="courseCard card"
      title="ZZZ1002"
    >
      <div
        class="card-body"
      >
        <div
          class="cardTitle card-title h5"
        >
          Sample2
        </div>
        <p
          class="courseSelectDetails card-text"
        >
          <span
            class="courseCodeText"
          >
            ZZZ1002
          </span>
          <span
            class="courseCredits"
          >
            4 Credits
          </span>
        </p>
        <div
          class="cardCompletedSubtitle card-subtitle h6"
        />
      </div>
    </div>
    <div
      class="courseCard card"
      title="ZZZ1001"
    >
      <div
        class="card-body"
      >
        <div
          class="cardTitle card-title h5"
        >
          Sample1
        </div>
        <p
          class="courseSelectDetails card-text"
        >
          <span
            class="courseCodeText"
          >
            ZZZ1001
          </span>
          <span
            class="courseCredits"
          >
            4 Credits
          </span>
        </p>
        <div
          class="cardCompletionFailedSubtitle card-subtitle h6"
        >
          Completed: N
        </div>
      </div>
    </div>
    <div
      class="courseCard card"
      title="ZZZ1003"
    >
      <div
        class="card-body"
      >
        <div
          class="cardTitle card-title h5"
        >
          Sample3
        </div>
        <p
          class="courseSelectDetails card-text"
        >
          <span
            class="courseCodeText"
          >
            ZZZ1003
          </span>
          <span
            class="courseCredits"
          >
            4 Credits
          </span>
        </p>
        <div
          class="cardCompletedSubtitle card-subtitle h6"
        >
          Completed: A
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`component: CourseCardList should render with filtered course list 1`] = `
<DocumentFragment>
  <div
    class="courseList card-columns"
  >
    <div
      class="courseCard card"
      title="ZZZ1001"
    >
      <div
        class="card-body"
      >
        <div
          class="cardTitle card-title h5"
        >
          Sample1
        </div>
        <p
          class="courseSelectDetails card-text"
        >
          <span
            class="courseCodeText"
          >
            ZZZ1001
          </span>
          <span
            class="courseCredits"
          >
            4 Credits
          </span>
        </p>
        <div
          class="cardCompletedSubtitle card-subtitle h6"
        />
      </div>
    </div>
    <div
      class="courseCard card"
      title="ZZZ1002"
    >
      <div
        class="card-body"
      >
        <div
          class="cardTitle card-title h5"
        >
          Sample2
        </div>
        <p
          class="courseSelectDetails card-text"
        >
          <span
            class="courseCodeText"
          >
            ZZZ1002
          </span>
          <span
            class="courseCredits"
          >
            4 Credits
          </span>
        </p>
        <div
          class="cardCompletedSubtitle card-subtitle h6"
        />
      </div>
    </div>
    <div
      class="courseCard card"
      title="ZZZ1003"
    >
      <div
        class="card-body"
      >
        <div
          class="cardTitle card-title h5"
        >
          Sample3
        </div>
        <p
          class="courseSelectDetails card-text"
        >
          <span
            class="courseCodeText"
          >
            ZZZ1003
          </span>
          <span
            class="courseCredits"
          >
            4 Credits
          </span>
        </p>
        <div
          class="cardCompletedSubtitle card-subtitle h6"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`component: CourseCardList should render without explicit props 1`] = `
<DocumentFragment>
  <div
    class="courseList card-columns"
  />
</DocumentFragment>
`;
